sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -3180,5 +3180,232 @@ def pipeline_component(input_list: Input[List[Artifact]]):
+            'exec-print-op-2')
+
+        # check tasks point to corrrect component spec reference
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op'].component_ref.name,
+            'comp-print-op')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].component_ref.name,
+            'comp-print-op-2')
+
+        # check information of resource spec is not lost
+        self.assertEqual(
+            4.0, pipeline_spec.deployment_spec.fields['executors'].struct_value
+            .fields['exec-print-op-2'].struct_value['container']
+            ['resources'].fields['cpuLimit'].number_value)
+        self.assertEqual(
+            15.032385536, pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields['exec-print-op-2'].struct_value['container']
+            ['resources'].fields['memoryLimit'].number_value)
+
+    def test_task_specific_configurations_would_dedupe(self):
+
+        @dsl.component
+        def print_op(message: str = 'default'):
+            print(message)
+
+        @dsl.pipeline()
+        def my_pipeline():
+            task1 = print_op()
+            task2 = print_op(message='different').set_display_name('name')
+
+        pipeline_spec = my_pipeline.pipeline_spec
+
+        self.assertEqual(len(pipeline_spec.components), 1)
+
+        # check one component spec and executor exist
+        self.assertIn('comp-print-op', pipeline_spec.components)
+        self.assertIn(
+            'exec-print-op', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+        self.assertNotIn('comp-print-op-2', pipeline_spec.components)
+        self.assertNotIn(
+            'exec-print-op-2', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+
+        # check tasks point to correct component spec
+
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op'].component_ref.name,
+            'comp-print-op')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].component_ref.name,
+            'comp-print-op')
+
+        # check task specific information is saved
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].inputs
+            .parameters['message'].runtime_value.constant.string_value,
+            'different')
+
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].task_info.name, 'name')
+
+    def test_pipeline_in_pipeline_would_dedupe(self):
+
+        @dsl.component
+        def print_op(msg: str):
+            print(msg)
+
+        @dsl.pipeline(name='graph-component')
+        def graph_component(msg: str):
+            print_op(msg=msg)
+
+        @dsl.pipeline(name='test-pipeline')
+        def my_pipeline():
+            task1 = graph_component(msg='hello')
+            task2 = graph_component(msg='hello')
```

Comment:
What happens if we reuse `print_op` again inside `my_pipeline`? Can we add this to the test case?

Reply:
Nothing particularly new happens, it adds a new task spec for `task3` `print_op` but still dedupes all `print_op ` component specs and executor_specs into one, would you still recommend i add it to the test case, if so what would you recommend i test for about it?

Reply:
I think it makes sense to test a root pipeline with the structure below. The interesting cases are where `comp-1` is deduped and the case where it is not. The same assertions you have for other tests make sense here too.

```txt
root-pipeline
    inner-pipeline
        comp-1
    comp-1
```